### PR TITLE
Add malloc-may-fail option to goto-check

### PR DESCRIPTION
**Pull request by xbauch**
And append an `assert(false)` property to every `malloc` call.

- [x] Each commit message has a non-empty body, explaining why the change was made.
- [ ] Methods or procedures I have added are documented, following the guidelines provided in CODING_STANDARD.md.
- [ ] The feature or user visible behaviour I have added or modified has been documented in the User Guide in doc/cprover-manual/
- [x] Regression or unit tests are included, or existing tests cover the modified code (in this case I have detailed which ones those are in the commit message).
- [ ] My commit message includes data points confirming performance improvements (if claimed).
- [x] My PR is restricted to a single feature or bugfix.
- [x] White-space or formatting changes outside the feature-related changed lines are in commits of their own.

### CODE REVIEW

**Comment by xbauch** [comment]
This may fix #4605.

**Review by allredj — COMMENTED** [review]
:heavy_check_mark:
Passed Diffblue compatibility checks (cbmc commit: effdd300e96d1165003a86502e8120e48b388db3).
Build URL: https://travis-ci.com/diffblue/test-gen/builds/143378007

**Review by martin-cs — COMMENTED** [review]
Why not just change the model of what malloc does?  That seems to keep all of the functionality in one place?

https://github.com/diffblue/cbmc/blob/develop/src/ansi-c/library/stdlib.c#L113

Maybe it is time to change https://github.com/diffblue/cbmc/blob/develop/src/ansi-c/library/stdlib.c#L115 specifically.

**Comment by xbauch** [comment]
> Why not just change the model of what malloc does?

I think the reasoning here is to avoid "false-positive" alarms with every `malloc` call by default, and only to be correctly reporting the potential of malloc to fail if requested.

**Comment by martin-cs** [comment]
> I think the reasoning here is to avoid "false-positive" alarms with
> every `malloc` call by default, and only to be correctly reporting
> the potential of malloc to fail if requested.

Then have a global as part of the model that allows it to be turned on
/ off / linked to a symbolic value during analysis.  It doesn't have to
be, and shouldn't be, hard-coded to one way or the other.

File: src/ansi-c/library/stdlib.c [hunk 1]
```diff
@@ -115,23 +115,31 @@ inline void *malloc(__CPROVER_size_t malloc_size)
   // realistically, malloc may return NULL,
   // and __CPROVER_allocate doesn't, but no one cares
   __CPROVER_HIDE:;
-  void *malloc_res;
-  malloc_res = __CPROVER_allocate(malloc_size, 0);
 
-  // make sure it's not recorded as deallocated
-  __CPROVER_deallocated=(malloc_res==__CPROVER_deallocated)?0:__CPROVER_deallocated;
+    __CPROVER_assert(
+      __CPROVER_malloc_may_fail == (__CPROVER_bool)0, "malloc may fail");
```
Comment (hannes-steffenhagen-diffblue) [inline src/ansi-c/library/stdlib.c]: ? Why is there an assert that this is false? I thought this would be more like `if (__CPROVER_malloc_may_fail && __VERIFIER_nondet___CPROVER_bool()) return NULL;`

Comment (xbauch) [reply src/ansi-c/library/stdlib.c]: Because conditionally returning `NULL` breaks too many things down the line in CBMC.

Comment (xbauch) [reply src/ansi-c/library/stdlib.c]: I take it back. Returning `NULL` seems to work here just fine.

File: src/ansi-c/library/stdlib.c [hunk 2]
```diff
@@ -115,23 +115,31 @@ inline void *malloc(__CPROVER_size_t malloc_size)
   // realistically, malloc may return NULL,
   // and __CPROVER_allocate doesn't, but no one cares
   __CPROVER_HIDE:;
-  void *malloc_res;
-  malloc_res = __CPROVER_allocate(malloc_size, 0);
 
-  // make sure it's not recorded as deallocated
-  __CPROVER_deallocated=(malloc_res==__CPROVER_deallocated)?0:__CPROVER_deallocated;
+    if(__CPROVER_malloc_may_fail && __VERIFIER_nondet___CPROVER_bool())
```
Comment (thk123) [inline src/ansi-c/library/stdlib.c]: :no_entry_sign: Comment above needs updating

File: regression/cbmc/Pointer_byte_extract5/main.i [hunk 3]
```diff
@@ -11,13 +11,13 @@ typedef union
 typedef struct
 {
   int Count;
-  Union List[1];
+  Union List[0];
```
Comment (thk123) [inline regression/cbmc/Pointer_byte_extract5/main.i]: :question: Why does this change? The change should go in a separate commit with a why

File: regression/cbmc/Pointer_byte_extract5/no-simplify.desc [hunk 4]
```diff
@@ -1,4 +1,4 @@
-CORE
+CORE broken-smt-backend
```
Comment (thk123) [inline regression/cbmc/Pointer_byte_extract5/no-simplify.desc]: :no_entry_sign: Please add an explanation / link to relevant issue as to why this is broken with smt-backend at the bottom of the desc file

File: regression/cbmc/Pointer_byte_extract5/test.desc [hunk 5]
```diff
@@ -4,7 +4,7 @@ main.i
 ^EXIT=10$
 ^SIGNAL=0$
 array\.List dynamic object upper bound in p->List\[2\]: FAILURE
-\*\* 1 of 11 failed 
+\*\* 1 of 12 failed
```
Comment (thk123) [inline regression/cbmc/Pointer_byte_extract5/test.desc]: :question: Why does this change? 
:pick: Probably best to set to `\d+` since number of goals isn't what is being tested

File: regression/cbmc/malloc-may-fail/test.desc [hunk 6]
```diff
@@ -0,0 +1,9 @@
+CORE
```
Comment (thk123) [inline regression/cbmc/malloc-may-fail/test.desc:1]: :no_entry_sign: Please add `.desc` file _without_ the option, that checks for `VERIFICATION SUCCESSFUL`

File: regression/goto-analyzer/constant_propagation_02/test.desc [hunk 7]
```diff
@@ -3,7 +3,7 @@ main.c
 --constants --simplify out.gb
 ^EXIT=0$
 ^SIGNAL=0$
-^Simplified:  assert: 1, assume: 0, goto: 1, assigns: 7, function calls: 0$
+^Simplified:  assert: 1, assume: 0, goto: 1, assigns: 8, function calls: 0$
```
Comment (thk123) [inline regression/goto-analyzer/constant_propagation_02/test.desc]: :pick: Please group these changes into a separate commit with why these numbers are changing

File: src/cbmc/cbmc_parse_options.cpp [hunk 8]
```diff
@@ -1041,6 +1041,7 @@ void cbmc_parse_optionst::help()
     "\n"
     "Backend options:\n"
     " --object-bits n              number of bits used for object addresses\n"
+    " --malloc-may-fail            enable failures for malloc calls\n"
```
Comment (thk123) [inline src/cbmc/cbmc_parse_options.cpp]: :bulb: I think it'd be better to be a little more explicit. 

```
allow malloc calls to return a null pointer
```

File: src/cbmc/cbmc_parse_options.h [hunk 9]
```diff
@@ -48,6 +48,7 @@ class optionst;
   "(document-subgoals)(outfile):(test-preprocessor)" \
   "D:I:(c89)(c99)(c11)(cpp98)(cpp03)(cpp11)" \
   "(object-bits):" \
+  "(malloc-may-fail)" \
```
Comment (thk123) [inline src/cbmc/cbmc_parse_options.h]: :grey_question: should this option be available in other C based tools, like goto-analyzer, or does that not really make sense?

File: src/util/config.cpp [hunk 10]
```diff
@@ -1093,6 +1093,11 @@ bool configt::set(const cmdlinet &cmdline)
     bv_encoding.is_object_bits_default = false;
   }
 
+  if(cmdline.isset("malloc-may-fail"))
```
Comment (thk123) [inline src/util/config.cpp]: :pick: ?

```suggestion
  ansi_c.malloc_may_fail = cmdline.isset("malloc-may-fail");
```

File: src/ansi-c/library/stdlib.c (hunk 2 above)
Comment (xbauch) [reply src/ansi-c/library/stdlib.c]: Updated.

File: regression/cbmc/Pointer_byte_extract5/main.i (hunk 3 above)
Comment (xbauch) [reply regression/cbmc/Pointer_byte_extract5/main.i]: This was undefined behaviour (and the changed in malloc exposed that CBMC does not really support that). It now has separate commit with explanation.

File: regression/cbmc/Pointer_byte_extract5/no-simplify.desc (hunk 4 above)
Comment (xbauch) [reply regression/cbmc/Pointer_byte_extract5/no-simplify.desc]: Added.

File: regression/cbmc/Pointer_byte_extract5/test.desc (hunk 5 above)
Comment (xbauch) [reply regression/cbmc/Pointer_byte_extract5/test.desc]: Explained and changed.

File: regression/cbmc/malloc-may-fail/test.desc (hunk 6 above)
Comment (xbauch) [reply regression/cbmc/malloc-may-fail/test.desc:1]: Added.

File: regression/goto-analyzer/constant_propagation_02/test.desc (hunk 7 above)
Comment (xbauch) [reply regression/goto-analyzer/constant_propagation_02/test.desc]: In separate commit now.

File: src/cbmc/cbmc_parse_options.cpp (hunk 8 above)
Comment (xbauch) [reply src/cbmc/cbmc_parse_options.cpp]: Changed.

File: src/cbmc/cbmc_parse_options.h (hunk 9 above)
Comment (xbauch) [reply src/cbmc/cbmc_parse_options.h]: Ambivalent.

File: src/util/config.cpp (hunk 10 above)
Comment (xbauch) [reply src/util/config.cpp]: Done.

File: regression/cbmc/malloc-may-fail/test.desc [hunk 11]
```diff
@@ -0,0 +1,9 @@
...
+main.c
+--malloc-may-fail
+^EXIT=10$
+^SIGNAL=0$
+^\[main.assertion.\d+\] line \d+ assertion p: FAILURE$
+^VERIFICATION FAILED$
+--
+^warning: ignoring
```
Comment (thk123) [inline regression/cbmc/malloc-may-fail/test.desc:9]: :pick: in future, we've found it helpful in test-gen to include a description explaining the purpose of the test below the line. for all tests. It helps when a test fails to understand why does it even exist. E.g

```
--
^warning: ignoring
--
Test that if --malloc-may-fail is enabled, that assert can return a null pointer
```

**Review by thk123 — APPROVED** [review]

**Review by smowton — APPROVED** [review]

File: src/ansi-c/library/stdlib.c [hunk 12]
```diff
@@ -115,6 +115,9 @@ inline void *malloc(__CPROVER_size_t malloc_size)
   // realistically, malloc may return NULL,
   // and __CPROVER_allocate doesn't, but no one cares
   __CPROVER_HIDE:;
+  if(__CPROVER_malloc_may_fail && __VERIFIER_nondet___CPROVER_bool())
+    return (void *)0;
```
Comment (tautschnig) [inline src/ansi-c/library/stdlib.c]: 1) It may be more consistent to just declare a local, uninitialised variable of type `__CPROVER_bool` to avoid the side-effect of a function call.
2) The non-determinism also needs to be introduced in `calloc`.
3) `realloc` now needs to handle the case of a failing `malloc`.

File: regression/cbmc/Pointer_byte_extract5/main.i (hunk 3 above)
Comment (tautschnig) [reply regression/cbmc/Pointer_byte_extract5/main.i]: I find it rather surprising that the changes in this PR expose the lack of support for this C90-hack. What exactly was going wrong here?

File: src/cbmc/cbmc_parse_options.cpp [hunk 13]
```diff
@@ -1041,6 +1041,7 @@ void cbmc_parse_optionst::help()
     "\n"
     "Backend options:\n"
     " --object-bits n              number of bits used for object addresses\n"
+    // NOLINTNEXTLINE(whitespace/line_length)
```
Comment (tautschnig) [inline src/cbmc/cbmc_parse_options.cpp]: The commit message says "will be squashed" so presumably that should happen ;-)

**Review by tautschnig — COMMENTED** [review]
The changes to extend support to calloc, realloc need to happen before this can be merged. The modified regression test I'd not insist as much on, but still I'd like to see clarification.

File: regression/cbmc/Pointer_byte_extract5/main.i (hunk 3 above)
Comment (xbauch) [reply regression/cbmc/Pointer_byte_extract5/main.i]: Why this particular PR exposes the problem is a bit more complicated. But the problem itself can be easily reproduced, e.g. (CMBC reports the assertion as failing):
```
#include <malloc.h>
#include <assert.h>

struct Foo
{
  int i;
  char data[1];
};

int main()
{
  struct Foo* foo = malloc(sizeof(struct Foo) + sizeof(char));
  foo->data[1]='b'; // set data[1]
  assert(foo->data[1]=='b'); // check data[1] -- should succeed

  return 0;
}
```

File: src/ansi-c/library/stdlib.c (hunk 12 above)
Comment (xbauch) [reply src/ansi-c/library/stdlib.c]: Add 1. No can do: using uninitialised variables doesn't even compile with the current flag set. I can initialise it to `nondet_bool`.

Comment (xbauch) [reply src/ansi-c/library/stdlib.c]: 2,3 are done.

File: src/cbmc/cbmc_parse_options.cpp (hunk 13 above)
Comment (xbauch) [reply src/cbmc/cbmc_parse_options.cpp]: Squashed.

**Comment by martin-cs** [comment]
There seems to be some duplication w.r.t. https://github.com/diffblue/cbmc/pull/5210  which is preferred?

**Comment by chrisr-diffblue** [comment]
@martin-cs I believe the difference between this PR and #5210 is around the behaviour they are modelling. #5210 is modelling the very specific case of what happens if the user code attempts to alloc a chunk of memory that is _larger than CBMC can represent_. Where as this PR is modelling the case where "the system" (e.g. OS) cannot satisfy a malloc request because of insufficient memory availability. This is a non-deterministic situation, so this PR models the situation where malloc may non-deterministically fail.

File: src/cbmc/cbmc_parse_options.cpp [hunk 14]
```diff
@@ -1041,6 +1041,8 @@ void cbmc_parse_optionst::help()
...
+    " --malloc-may-fail            allow malloc calls to return a null pointer\n"
```
Comment (chrisr-diffblue) [inline src/cbmc/cbmc_parse_options.cpp]: Similar to #5210, should this help message move up into the "Program instrumentation options" section?

**Review by chrisr-diffblue — COMMENTED** [review]
Looks pretty good, but would also be super nice if you could extend the user docs at the same time ;-)

File: src/cbmc/cbmc_parse_options.cpp (hunk 14 above)
Comment (xbauch) [reply src/cbmc/cbmc_parse_options.cpp]: Moved to the instrumentation section + the documentation was also added.

File: src/ansi-c/library/stdlib.c [hunk 15]
```diff
@@ -112,50 +137,55 @@ __CPROVER_bool __VERIFIER_nondet___CPROVER_bool();
...
+    __CPROVER_malloc_failure_mode ==
+    __CPROVER_malloc_failure_mode_assert_then_assume)
+  {
+    __CPROVER_assert(
+      malloc_size <= __CPROVER_max_malloc_size, "max allocation size exceeded");
+    __CPROVER_assume(malloc_size <= __CPROVER_max_malloc_size);
+
+    __CPROVER_bool should_malloc_fail = __VERIFIER_nondet___CPROVER_bool();
+    __CPROVER_assert(
+      !__CPROVER_malloc_may_fail || !should_malloc_fail,
```
Comment (hannes-steffenhagen-diffblue) [inline src/ansi-c/library/stdlib.c:164]: I don’t think it makes a lot of sense to have the assertion here in this form, this will _always_ fail the way it is written.

**Review by hannes-steffenhagen-diffblue — COMMENTED** [review]
This should have an implementation for `realloc` as well, also I don’t think we should mix up the cases where malloc fails due to some nondeterministic condition and when it fails because we reach cprover internal limits.

**Review by chrisr-diffblue — APPROVED** [review]

**Review by martin-cs — APPROVED** [review]
I like this approach a lot more than the previous one.
